include/pybind11/numpy.h

```diff
@@ -42,6 +44,30 @@ class array; // Forward declaration
 
 PYBIND11_NAMESPACE_BEGIN(detail)
 
+// Main author of this class: jbms@
+template <typename T>
+class LazyInitializeAtLeastOnceDestroyNever {
+public:
+    template <typename Initialize>
+    T &Get(Initialize &&initialize) {
+        if (!initialized_) {
+            assert(PyGILState_Check());
+            // Multiple threads may run this concurrently, but that is fine.
```

Comment:
This comment is a bit misleading. Multiple threads may call `Get`, but multiple threads definitely mustn't get to this point! E.g. `initialized_` is not an atomic variable and it'd be a race to access it concurrently.

What is true is that multiple threads may possibly reach the inside of `initialize`, but that's a separate matter.

Reply:
I'm uncertain I understand the two points correctly.

My understanding:

From getting burned in the past, I've developed the strong belief that multiple threads do in fact get concurrently into the `if (! initialized_)` branch.

If the first thread calls back into the Python C API (e.g. for a Python import), Python can and does in the general case release the GIL. That gives other threads the opportunity to also pass the `if (! initialized_)` test, so the Python `initialize()` really can run multiple times.

But after the `initialize()` call we are sure to have the GIL again, therefore there is no race flipping `initialized_` from `false` to `true`. One lucky thread gets there first. (All other threads that made it to the `initialize()` call will write the `true` again, but each under the GIL.) Only then will other threads no longer get into the `if (! initialized_)` branch.

Would it help to change the comment to like this?

```
// Multiple threads may reach `initialize()` concurrently, but each holding the GIL at that time.
```
________

Fundamentally, the "at least once" aspect here isn't great. "Guaranteed once" would be better.

My thinking:

* Let's get the "at least once" solution stable. It almost seems to be there, and it solves the original deadlock problem.

* Then let's look at it again (this PR maybe) to see what it takes to achieve a "guaranteed once" implementation (right here in pybind11, within the limitations of C++11).

Reply:
Sorry, I think when I said "multiple threads can call `Get`" that's not entirely right. By assuming that `Get` is only called with the GIL held, we effectively say that only one thread can call `Get` at a time. What I meant, but didn't say right, was that "mulitple threads can attept to lock the GIL and call `Get`".

> From getting burned in the past, I've developed the strong belief that multiple threads do in fact get concurrently into the `if (! initialized_)` branch.

I'm not sure how this would happen, and if it did, it'd definitely be a bug, since that's a race.

> If the first thread calls back into the Python C API (e.g. for a Python import), Python can and does in the general case release the GIL. That gives other threads the opportunity to also pass the `if (! initialized_)` test, so the Python `initialize()` really can run multiple times.

OK, but that is not "concurrently". Only the thread holding the GIL can be accessing `initialized_`. In order for one thread to release the GIL it must already have completed this access. (And here "completed" can be made precise in the sense of the memory model.)

I don't think "`initialize()` really can run multiple times" is a precise enough statement to allow for a meaningful interpretation. It is certainly true that multiple threads can exectue inside `initialize`, but it is (or definitely _should_) _not_ be true that multiple threads reach the expression `initialize()` on L56 "concurrently", in the sense of "without sequencing". If one thread is inside `initialize` and another thread reaches L56, then the former thread must be blocked.

Reply:
Maybe we can say something more concrete and specific to this situation:

```
// It is possible that multiple threads execute `Get` with `initialized_` still being false,
// and thus proceed to execute `initialize()`, which requires that `initialize` internally
// releases and reacquires the GIL. We accept this, and expect the operation to be
// both idempotent and cheap.

Reply:
Sorry I appear to be mixing up terms, which is critical here.

What is the correct term? (I want to edit-correct my comment.)

E.g.

> I've developed the strong belief that multiple threads do in fact get ~concurrently~ CORRECT_TERM_HERE into the if (! initialized_) branch.

I want to say what you are saying.

Although all I originally wanted: leave a meaningful, terse, and correct comment in the code, just enough to give readers a good clue to follow up on in case they want to fully understand.

Reply:
Oh, I'm seeing your suggested comment only now. Will adopt.

Reply:
Done: 1ce27154d655dd768ddd855394f30969c476b9fb